### PR TITLE
Add a method for Numerical Surface: Can get both coordinates and atom type

### PR DESCRIPTION
**Pull request by Luckick**
getAllPointswithAtomType(): Will return an ArrayList, where each
element will include both Coordinates and Atom Type by
Atom.getAtomicNumber().

### CODE REVIEW

File: descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/Point_Type.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2004-2007  The Chemistry Development Kit (CDK) project
```
Comment (egonw) [inline descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/Point_Type.java]: It this file a copy of an existing CDK file? If not, please list here 2019 and your own name or that of you employer.

File: descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/NumericalSurface.java [hunk 2]
```diff
@@ -189,6 +190,29 @@ private void init() {
         }
         return (ret);
     }
+    
+    public ArrayList<Point_Type> getAllPointswithAtomType() {
```
Comment (egonw) [inline descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/NumericalSurface.java]: 1. Can you please add a JUnit test case for this new methods too?
2. Please provide JavaDoc to describe the functionality.
3. I think the return type can be just `List<Point_Type>`

File: descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/Point_Type.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package org.openscience.cdk.geometry.surface;
+
+import javax.vecmath.Point3d;
+
+
+public class Point_Type
```
Comment (egonw) [inline descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/Point_Type.java]: I personally prefer it to have no hyphen in the class name.
Also, please add JavaDoc to all public classes and methods.

Comment (johnmay) [reply descriptor/qsarmolecular/src/main/java/org/openscience/cdk/geometry/surface/Point_Type.java]: No new class needed, this is a simple pair, the method return type should be ``List<Map.Entry<IAtom,Point3d>>``.

**Comment by johnmay** [comment]
Thanks, looks simple. Few things to tidy up, as I say in another comment the return type should be ``List<Map.Entry<IAtom,Point3d>>`` using ``new AbstractMap.SimpleEntry<>(atom, surfpoint)``. Likewise what raw types in the method, they may have already been there though.

**Comment by Luckick** [comment]
Thank both of you! 
I will think about the way that you mentioned and commit latter.

**Comment by johnmay** [comment]
I'll try a take a look this evening.

**Comment by Luckick** [comment]
Great!
Another reminder is that since the tessellation process focus on triangles and count every vertexes of triangles into surface points, there are some repeated points due to triangles share vertexes.

**Comment by johnmay** [comment]
Just to clarify the AdapativeNumericalSurface is a new method.

**Comment by johnmay** [comment]
Doesn't look like it, have cleaned up #559. Please check it does everything you need. I replaced the Point_Type with a Map<IAtom,List<Point3d>> as that made more sense.

**Comment by johnmay** [comment]
@egonw @Luckick I think this can be closed. A previous #533 exposed the required functionality from the NumericalSurface allowing the pairing of IAtoms and Points on the surface. We could add an additional "flat-map" which unpacks the atom <=> point relationship 1-to-1 as follows: 

```java
 public List<Map.Entry<IAtom, Point3d>> getAtomSurfaceFlatMap() {
    List<Map.Entry<IAtom,Point3d>> res = new ArrayList<>();
    for (int i = 0; i < this.surfPoints.length; i++) {
        if (!this.surfPoints[i].isEmpty()) {
            final IAtom atom = this.atoms[i];
            this.surfPoints[i].forEach(p -> res.add(new SimpleImmutableEntry<>(atom, p)));
        }
    }
    return res;
}
```

This is more similar to the Point_Type in this PR but I think the Atom -> List<Point> is better since it's what is stored internally.

I've just gone through the 3000 file diff to see what the actual change was.

I find the following in NumericalSurface: 
Screenshot 2022-02-05 at 13 27 23

Since Point_Type does implement equals/hashCode the contains() does is doing a reference comparison and will always be false.

and then another Tessellate:
Screenshot 2022-02-05 at 13 27 39

This is fine (sort of) but I would make it another method. I said it's sort of fine because de-duplicate floating point is hard, probably what is needed is some threshold where we consider overlapping points to be identical. Since this is outside the scope of the original requirement and a simple two-liner:

```java
public Point3d[] getTessAsPoint3dsUniq() {
    Set<Point3d> set = new HashSet<>(Arrays.asList(getTessAsPoint3ds()));
    return set.toArray(new Point3d[0]);
}
``` 

I would propose closing this and resubmitting any other changes that are needed.

**Comment by johnmay** [comment]
@egonw okay to close? (reasons above)

**Comment by egonw** [comment]
No objections.

**Pull request closed without merging**